proxy/proxy.go

```diff
@@ -280,13 +280,17 @@ func (proxy *Proxy) Listen() []net.Listener {
 	return listeners
 }
 
-func (proxy *Proxy) Serve(listeners []net.Listener) {
+func (proxy *Proxy) Serve(listeners []net.Listener, ready func()) {
 	errs := make(chan error)
 	for _, listener := range listeners {
 		go func(listener net.Listener) {
 			errs <- (&http.Server{Handler: proxy}).Serve(listener)
 		}(listener)
 	}
+	// It would be better if we could delay calling Done() until all
+	// the listeners are ready, but it doesn't seem to be possible to
+	// hook the right point in http.Server
+	ready()
 	for range listeners {
 		err := <-errs
 		if err != nil {
```

Comment:
Not caused by this change, but I noticed that, at least [according to the docs](https://golang.org/pkg/net/http/#Server.Serve) this condition should always be met, leading to potentially confusing log messages:
> Serve always returns a non-nil error. 
> After Shutdown or Close, the returned error is ErrServerClosed.

It should probably check if the error returned is `ErrServerClosed` instead.